internal/auth/client_kerberos.go

```diff
@@ -0,0 +1,477 @@
+	// Request is the action that's being requested (50: SSH_MSG_USERAUTH_REQUEST)
+	Request byte
+	// UserName is the username that the user requests to log in as
+	UserName string
+	// Service is the service being used ('ssh-connection')
+	Service string
+	// Method is the authentication method in use ('gssapi-with-mic')
+	Method string
+}
+
+func (mic *GSSAPIMicField) unmarshal(b []byte) error {
+	err := ssh.Unmarshal(b, mic)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kerberosAuthContext) VerifyMIC(micField []byte, micToken []byte) error {
+	var t gssapi.MICToken
+	err := t.Unmarshal(micToken, false)
+	if err != nil {
+		return err
+	}
+	t.Payload = micField
+	verified, err := t.Verify(k.key, keyusage.GSSAPI_INITIATOR_SIGN)
+	if err != nil {
+		return err
+	}
+	if !verified {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
+		)
+	}
+
+	// MIC is verified, but need to ensure usernames match
+	var field GSSAPIMicField
+	err = field.unmarshal(micField)
+	if err != nil {
+		return err
+	}
+
+	if field.Request != SSH_MSG_USERAUTH_REQUEST || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
+		)
+	}
+
+	if k.client.config.EnforceUsername && field.UserName != k.username {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
+		)
+	}
+
+	k.success = true
+
+	return nil
+}
+
+func (k *kerberosAuthContext) DeleteSecContext() error {
+	return nil
+}
+
+func (k *kerberosAuthContext) tryAllowLogin(username  string) (bool ,error) {
+	authzRequest := auth.AuthorizationRequest{
+		PrincipalUsername: k.username,
+		LoginUsername: username,
+		RemoteAddress: k.remoteAddr.String(),
+		ConnectionID: k.connectionId,
+	}
+	authzResponse := &auth.ResponseBody{}
+
+	statusCode, err := k.client.http.Post("/authz", authzRequest, authzResponse)
+	if err != nil {
+		k.err = err
+		k.success = false
+		return false, err
+	}
+	if statusCode != 200 {
+		k.err = message.Wrap(
+			err,
+			message.EAuthKerberosBackendError,
+			"Non 200 error code from authz server: %d",
+			statusCode,
+		)
+		k.success = false
+		return false, k.err
+	}
+
+	if authzResponse.Success {
+		k.metadata = authzResponse.Metadata
+		return true, nil
+	}
+
+	k.success = false
+	return false, nil
+}
+
+func (k *kerberosAuthContext) AllowLogin(username string) error {
+	if !k.Success() {
+		return k.Error()
+	}
+	if k.client.config.EnforceUsername && k.username != username {
+		return message.NewMessage(
+			message.EAuthKerberosUsernameDoesNotMatch,
+			"User %s tried to log in as %s but enforceUsername was enabled",
+			k.username,
+			username,
+		)
+	}
+	if !k.client.config.Authz.Enable {
+		// Allow all since authz server is disabled
+		return nil
+	}
+
+	retrier := retry.New(
```

Comment:
Can we move adding retries to a separate PR and do it the manual way here? I'd rather introduce retries globally in a single PR, that would make it easier to add customizations such as logging.